Update Theme articles with Application Theme (#41)
vaadin/docs: articles/themes/application-theme.asciidoc

---
title: Application Theme
order: 25
layout: page
---

= Application Theme
ifdef::web[]
endif::web[]

Application Theme is the standard way to organize themes in a Vaadin application, making it easier to switch between themes and allow packaging a theme to reuse in other Vaadin applications.

== Define application themes

> Figure 1 illutrates the contents of a theme:

Reviewer: ```suggestion
Figure 1 illustrates the contents of a theme:
```

NOTE: The static resource must be available from the application theme folder for it to be handled, else it should be available with the given URL.

=== Global Styles

Application-specific styling across application views and application-specific components are defined in CSS files under the theme folder.

For example, <<css-custom-properties#,CSS Custom Properties>> or `@font-face` can be defined in a `global.css` file.
See <<styling-applications#, Styling Applications>>.

NOTE: Global styles can override <<theme-variants#global-variants, global variants>> styles.
For example, `Lumo.DARK` variant defines a `--lumo-base-color` which can be overridden by `global.css` file.

=== Theme Configuration	

==== Assets	

Assets like fonts and images from an npm dependency can be used in a theme after mapping them from the dependency to local URIs by adding an `assets` block in `theme.json`.	

[source, JSON]	
----	
"assets": {	
  "@fortawesome/fontawesome-free": {
    "svgs/regular/**": "fortawesome/icons"
  }
}	
----	

A style sheet in the theme can use the assets by the local path defined in `theme.json`.	

[source, CSS]	
----	

> background-image: url('fortawesome/icons/address-book.svg');	

Reviewer 2: ```suggestion
background-image: url('./fortawesome/icons/address-book.svg');	
```

Reviewer 2: Except I'm not certain that this will work so it might be that it needs `VAADIN/static/fortawesome/icons/...` should be added as a test that how do node assets work inside the css files.

At the moment we only test like: `        Image snowFlake = new Image("VAADIN/static/fortawesome/icons/snowflake.svg", "snowflake");`

Reviewer: Is it on purpose “fort” and not “font”?

Author: It's "fort". I checked the library: https://www.npmjs.com/package/@fortawesome/fontawesome-free .

Author: Added `VAADIN/static/`.

Reviewer 2: It's FontAwesome by [FortAwesome ](https://github.com/FortAwesome)
I just set it in the feature test to 'fortawesome/icons' as in they are icons from fortawesome, could be something else also.

---
title: Application Theme
order: 25
layout: page
---

= Application Theme

Application Theme is the standard way to organize themes in a Vaadin application, making it easier to switch between themes and allow packaging a theme to reuse in other Vaadin applications.

== Define Application Themes

Figure 1 illustrates the contents of a theme:

.Contents of an application theme
image::images/content-of-application-theme.png[width="70%"]

> === Theme Folder

Reviewer 2: Perhaps explain the folders as in the image e.g.
`theme` should be located under `./frontend` and contains all application themes in theme name folders.
`my-theme` contains all files associated with the theme my-theme
`components` is for component specific themes, style sheets need to match the component tag name.
`theme.json` theme configuration for assets etc.
`images` and `webfonts` are folders for static resources
`global.css` contains theme changes that affect the whole page.

Or else move the figure after the explanation of the different parts with the topic sample theme setup...

Author: I updated contents in the picture. I'd prefer to put it before the detail explanation so users can have a quick visualization.

Application themes are placed under the `frontend/theme/` folder.
This location is fixed.
Each theme is in a separate folder, the folder name is also the theme name.

=== Component Styles

Component-specific style sheets are placed under the `components` folder.
The file name must match the component name that you want to style.
See <<styling-components#,Styling Components>>.

=== Static Resources

You can create folders for static resources, such as images or fonts.
When using in a style sheet, the resource location is relative to `frontend/theme/` folder.

> For example, in Figure 1, `global.css` file can have `background: url("theme/my-theme/images/bg.jpg");`

Reviewer 2: https://github.com/vaadin/flow/pull/9451 will change the url format to be `url('./images/bg.jpg');`

Author: Fixed. "the resource location is relative to the CSS file location". (or is it relative to the app theme folder?)

Reviewer 2: Realtive to the css file, but should still be inside the theme folder.
e.g. 
theme
|--> out-scope.png
|  |--> my-theme
|  |--> global.css
|  |--> img
|  |   |--> bg.jpg
|  |--> components
|     |--> vaadin-button.css

To use the bg image for global.css it would be `url('./img/bg.jpg')` and for button `url('../img/bg.jpg')`
Where as trying to get `out-scope.png` with either `url('../out-scope.png')` for global or `url('../../out-scope.png')`  for baadin-button would result in no rewrite and would fail on the server if it wouldn't be externally served.

Author: Fixed.

---
title: Application Theme
order: 25
layout: page
---

= Application Theme

Application Theme is the standard way to organize themes in a Vaadin application, making it easier to switch between themes and allow packaging a theme to reuse in other Vaadin applications.

== Define Application Themes

Figure 1 illustrates a sample theme setup for "my-theme" application theme.

.A sample application theme "my-theme"
image::images/content-of-application-theme.png[width="80%"]

=== Application Theme Folder

> Application themes are placed under the `frontend/theme/` folder.

Reviewer 3: First approach is better IMO. Thanks!

This location is fixed.
Each theme is in a separate folder, the folder name is also the theme name.

=== Component Styles

Component-specific style sheets are placed under the `components` folder.
The file name must match the component name that you want to style.
See <<styling-components#,Styling Components>>.

=== Static Resources

You can create folders for static resources, such as images or fonts.
When using in a style sheet, the resource location is relative to the CSS file location.

For example, in Figure 1, `global.css` file can have `background: url('./images/bg.jpg');`

> WARNING: The resource and the CSS file must be in the same application theme.

Author: Also need to mention external URL won't be touch.

=== Global Styles

Application-specific styling across application views and application-specific components are defined in CSS files under the theme folder.

For example, <<css-custom-properties#,CSS custom properties>> or `@font-face` can be defined in a `global.css` file.
See <<styling-applications#,Styling Applications>>.

=== Theme Configuration

> ==== Assets

Reviewer 2: Assets shoul be in another PR that is merged after we have implemented the feature.

Author: Removed.

Assets like fonts and images from an npm dependency can be used in the theme after mapping them from the dependency to local URIs, by adding an `assets` block in `theme.json`.

[source, JSON]
----
"assets": {
  "@foo/bar": {
    "images/**": "foobar/images"
  }
}
----

A style sheet in the theme can use the assets using the local path defined in `theme.json`.

[source, CSS]
----
background-image: url('foobar/images/logo.svg');
----

> ==== Extend an Application Theme

Reviewer 2: This should be removed and added in another PR after we have the parent theme feature implemented.

Author: Removed.

An application theme can extend another one by adding a `parent` property to theme.json.
For example, to extend a theme called "acme-base":

[source, JSON]
----
{
  "parent": "acme-base"
}
----

==== Use at-rules for Components in non-Vaadin Applications

> To make at-rules work for components embedded into non-Vaadin applications (for example <<../flow/embedding-flow-applications/tutorial-webcomponent-intro,WebComponentExporter>>), add a file called `document.css` to the theme folder, and write at-rules there.

Reviewer 3: Couldn't open this link in IntellJIDEA preview, it says the `file doesn't resolve`. And for some other links, for instance, `<<css-custom-properties#,CSS custom properties>>` it says `Anchor doesn't resolve`. Please double check.  But it can be only me who has it, or IntelliJIDEA preview issue.

CSS in this file is applied to the document root rather than the page body or shadow root, to support at-rules that don't work inside shadow roots.

At-rules in npm dependencies are declared in `theme.json` using `importCss` property:
[source, JSON]
----
{

>   "importCss": ["@foo/bar/baz.css"] 

Reviewer 3: Ok

}
----

NOTE: It is highly recommended to use https://vaadin.com/api/platform/com/vaadin/flow/component/dependency/NpmPackage.html[`@NpmPackage`] for installing npm dependencies so they won't be lost in case you accidentally remove `package.json`.

== Use application themes

In server-side views (Java), use the [classname]#@Theme# annotation to apply an application theme.

> You must set the annotation on an https://vaadin.com/api/platform/com/vaadin/flow/component/page/AppShellConfigurator.html[AppShellConfigurator] class.

Reviewer 3: Exception is being thrown otherwise?

== Define Application Themes

Figure 1 illustrates a sample theme setup for "my-theme" application theme.

.A sample application theme "my-theme"
image::images/content-of-application-theme.png[width="80%"]

=== Application Theme Folder
Application themes are placed under the `frontend/theme/` folder of a Vaadin project's root location.
This location is fixed.
Each theme is in a separate folder, the folder name is also the theme name.

=== Component Styles

Component-specific style sheets are placed under the `components` folder.
The file name must match the component name that you want to style.
See <<styling-components#, Styling Components>>.

NOTE: Component styles can override <<theme-variants#component-variants, component variants>> styles. For example, styles in `vaadin-button.css` file can override styles from `ButtonVariant.LUMO_ERROR` variant of a Vaadin Button component.

=== Static Resources

You can create folders for static resources, such as images or fonts.
When using in a style sheet, the resource location is relative to the CSS file location.

For example, in Figure 1, `global.css` file can have `background: url('./images/bg.jpg');`

> WARNING: The static resource should be in the same theme with the CSS file. Otherwise, it will be served as external resources by the web server instead of webpack.

Author: Thanks. Updated.

---
title: Application Theme
order: 25
layout: page
---

= Application Theme

Application Theme is the standard way to organize themes in a Vaadin application, making it easier to switch between themes and allow packaging a theme to reuse in other Vaadin applications.

== Define Application Themes

Figure 1 illustrates a sample theme setup for "my-theme" application theme.

.A sample application theme "my-theme"
image::images/content-of-application-theme.png[width="80%"]

=== Application Theme Folder
Application themes are placed under the `frontend/theme/` folder of a Vaadin project's root location.

> This location is fixed.

Reviewer 2: ```suggestion
The theme folder location is not configurable.
```
Might sound better.

> Each theme is in a separate folder, the folder name is also the theme name.

Reviewer 2: Should we have a couple examples here like 
- For `@Theme("alpaca")` the files should be in `frontend/theme/alpaca` 
- For `@Theme("frost-bite")`  the files should be in `frontend/theme/frost-bite`

Author: Added.

=== Component Styles

Component-specific style sheets are placed under the `components` folder.

> The file name must match the component name that you want to style.

Reviewer 2: ```suggestion
The file name must match the tag name of the component that you want to style.
```

See <<styling-components#, Styling Components>>.

> NOTE: Component styles can override <<theme-variants#component-variants, component variants>> styles. For example, styles in `vaadin-button.css` file can override styles from `ButtonVariant.LUMO_ERROR` variant of a Vaadin Button component.

Reviewer: ```suggestion
NOTE: Component styles can override <<theme-variants#component-variants, component variants>> styles.
For example, styles in `vaadin-button.css` file can override styles from `ButtonVariant.LUMO_ERROR` variant of a Vaadin Button component.
```

=== Static Resources

You can create folders for static resources, such as images or fonts.
When using in a style sheet, the resource location is relative to the CSS file location.

For example, in Figure 1, `global.css` file can have `background: url('./images/bg.jpg');`

NOTE: The static resource must be available from the application theme folder for it to be handled, else it should be available with the given URL.

=== Global Styles

Application-specific styling across application views and application-specific components are defined in CSS files under the theme folder.

For example, <<css-custom-properties#,CSS Custom Properties>> or `@font-face` can be defined in a `global.css` file.
See <<styling-applications#, Styling Applications>>.

> NOTE: Global styles can override <<theme-variants#global-variants, global variants>> styles. For example, `Lumo.DARK` variant defines a `--lumo-base-color` which can be overridden by `global.css` file.

Reviewer: ```suggestion
NOTE: Global styles can override <<theme-variants#global-variants, global variants>> styles.
For example, `Lumo.DARK` variant defines a `--lumo-base-color` which can be overridden by `global.css` file.
```

> ==== Use at-rules for Components in non-Vaadin Applications

Reviewer 2: Embedded components do not yet have the application theme support. This could also be moved to a new PR that the adds the information when the feature has been released.

Author: Removed.

For example, in Figure 1, `global.css` file can have `background: url('./images/bg.jpg');`

> WARNING: The resource and the CSS file must be in the same applicatin theme.

Reviewer 2: ```suggestion
NOTE: The resource and the CSS file must be in the same application theme to be handled, else it should be available with the given URL.
```
